### PR TITLE
enable training using chat templat

### PR DESCRIPTION
**Pull request by mina-parham**

### CODE REVIEW

File: transformerlab/plugins/llama_trainer/main.py [hunk 1]
```diff
@@ -49,20 +48,22 @@ def find_lora_target_modules(model, keyword="proj"):
...
 
-    # Setup template for formatting
-    print("Formatting template: " + tlab_trainer.params.formatting_template)
-    template = jinja_environment.from_string(tlab_trainer.params.formatting_template)
-
-    def format_instruction(mapping):
-        return template.render(mapping)
+    # Prepare datasets into JSONL files
+    data_directory = f"{WORKSPACE_DIR}/plugins/llama_trainer/data"
+    if not os.path.exists(data_directory):
```
[the remark this answers is not in the document]
Comment (deep1401) [reply transformerlab/plugins/llama_trainer/main.py:54]: Just a note here that apply_chat_template would only work for instruct models so for a base model being trained, we might have to do some sort of conditional check and still apply?

File: transformerlab/plugins/llama_trainer/main.py [hunk 2]
```diff
@@ -196,7 +197,7 @@ def format_instruction(mapping):
             save_strategy="epoch",
             learning_rate=learning_rate,
             bf16=True,
-            tf32=True,
+            tf32=False, # T4 GPUs do not support tf32
```
[the remark this answers is not in the document]
Comment (mina-parham) [reply transformerlab/plugins/llama_trainer/main.py:200]: It works on modern nvidia gpus which speeds up the most common calculations without losing much accuracy

File: transformerlab/plugins/llama_trainer/main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mina-parham) [reply transformerlab/plugins/llama_trainer/main.py]: we're handling this condition on the frontend, the chat template toggle is only shown to the user if the model has a non empty chat_template in the config file, so we only have this feature for instruct models

**Review by deep1401 — CHANGES_REQUESTED** [review]

File: transformerlab/plugins/llama_trainer/main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (deep1401) [reply transformerlab/plugins/llama_trainer/main.py]: I added this comment in my review again but maybe we should also have a condition on the backend so that if we do a remote run anytime then the plugin would still work

File: transformerlab/plugin_sdk/transformerlab/utils.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                if first_line:
+                    parsed = json.loads(first_line)
+                    print(f"Example from {split_name} split:")
+                    print(parsed.get("text", first_line))
+                else:
+                    print(f"Example from {split_name} split: file is empty.")
+        except Exception as e:
+            print(f"Error reading example from {output_file}: {e}")
+
+def format_template(
```
[the remark this answers is not in the document]
Comment (deep1401) [reply transformerlab/plugin_sdk/transformerlab/utils.py:53]: Sorry about the double comments and having to track this in 2 places, but my point was that we should also raise an error in the format_template function so that we don't run into an issue when we try to do a run from a recipe or something like that where things are pre-specified or even while doing a remote run.

File: transformerlab/plugins/mlx_lora_trainer/main.py [hunk 4]
```diff
@@ -4,25 +4,23 @@
...
 import subprocess
 import os
-from jinja2 import Environment
 
 
 # Import tlab_trainer from the SDK
 # from transformerlab.tlab_decorators import tlab_trainer
 from transformerlab.sdk.v1.train import tlab_trainer
 from transformerlab.plugin import WORKSPACE_DIR
+from transformerlab.utils import prepare_dataset_files
```
Comment (deep1401) [inline transformerlab/plugins/mlx_lora_trainer/main.py:17]: You might need to change this one now?

File: transformerlab/plugins/llama_trainer_multi_gpu/main.py [hunk 5]
```diff
@@ -3,9 +3,12 @@
 import time
 from random import randrange
 import torch.nn as nn
+from functools import partial
+
 
 from transformerlab.sdk.v1.train import tlab_trainer
 from transformerlab.plugin import WORKSPACE_DIR, get_python_executable
+from transformerlab.utils import format_template
```
Comment (deep1401) [inline transformerlab/plugins/llama_trainer_multi_gpu/main.py:11]: This one as well would need to be changed?

**Review by deep1401 — CHANGES_REQUESTED** [review]

**Review by deep1401 — APPROVED** [review]
Tested this on all 3 plugins and nothing breaks now

File: transformerlab/plugin_sdk/transformerlab/plugin.py [hunk 6]
```diff
@@ -590,3 +593,75 @@ def setup_model_worker_logger(name: str = "transformerlab") -> logging.Logger:
...
+    """ Prepares dataset files for training by formatting each example according to the provided template."""
+    tokenizer = None
+    if chat_template:
+        tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
+
+    for split_name in datasets:
+        dataset_split = datasets[split_name]
+        print(f"Processing {split_name} dataset with {len(dataset_split)} examples.")
+
+        output_file = f"{data_directory}/{split_name}.jsonl"
```
Comment (dadmobile) [inline transformerlab/plugin_sdk/transformerlab/plugin.py:614]: Tiny nitpick: We're trying to use os.path.join() when making filenames if possible. Although now that we are fully linux it possibly matters less.

**Review by dadmobile — CHANGES_REQUESTED** [review]

**Review by dadmobile — APPROVED** [review]
